### DISCUSSION TITLE
A few beginners questions

### DISCUSSION THREAD

**Discussion by bartbutenaers** [QuorumDMS/ftp-srv#310] [Q&A] [answered]
Hi,
Thanks for sharing this nice library!!

I have a few beginners questions.  Things that are not clear to me after reading the readme page:

1. There is a nice example to detect when the server is starting:
   ```
   ftpServer.listen().then(() => { 
      console.log('Ftp server is starting...')
   });
   ```
   However I should also be aware when the server is completely started.  Is there an event to accomplish that?

2. There are 3 examples of how to listen for command events, for example:
   ```
   connection.on('RNTO', (error, fileName) => { ... });
   ```
   Are there always two parameters, also for the events of commands not being listed on the readme page?

3. I would like to keep track of the number of connected clients.  I assume I can increase the clientCount when a succesful login event occurs?  But is there also a event that I can use to decrease the clientCount?  Because I dont' see an event being emitted [where](https://github.com/QuorumDMS/ftp-srv/blob/f59857e34aae906ad87264e13b6604467d3d4d6c/src/connection.js#L72) the connection is being closed.

Thanks!!
Bart

**Comment by matt-forster** [comment, marked as answer]
Hi, @bartbutenaers. I’ve converted this issue to a discussion because its just questions. I’ll do my best to answer here;

1. After that first listen promise completes, the server is started. That is confusing wording in the example.
2. The events that are meant for use are documented here: https://github.com/QuorumDMS/ftp-srv#events. You’ll notice they do not all of the same parameters for the handler function
3. I think you are on the right track - but you are right, there is no real good way to tell how many users are connected at the moment. I think an event would be appropriate for this, we are definitely open to contributions on this front!

**Comment by bartbutenaers** [comment]
Hey @matt-forster,
Apologies!  I had completely overlooked that you had activated the Github discussions...

1. Ok, that is clear.  Thanks!
2. Yes indeed I had seen that - in the 4 handler examples on the readme page - they don't all have the same parameters in the handler functions.  I will try to rephrase my question better: suppose I want to write a handler function for DELE command (for which there is no example on the readme page).   What is the easiest way for me to determine the parameter list, because at first sight I don't think that I can find this info in [this](https://github.com/QuorumDMS/ftp-srv/blob/main/src/commands/registration/dele.js) code snippet?
3. Ok then I will need to have a look.  Might take a bit longer, because I am helping my sons these days with their examinations ;-)
4. I hope it is not a problem if I add a new beginner question in this same discussion?  
   When I pass an URL without IP address (e.g. `ftp://:21`) via the options in the constructor, then I get this error in my console log:
   ```
   Unhandled rejection Error: getaddrinfo ENOTFOUND undefined
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:71:26)
    at GetAddrInfoReqWrap.callbackTrampoline (node:internal/async_hooks:130:17)
   ```
   BTW it is just a hypothetical use case to simulate an error.
   I was wondering if an error occurs, can I capture that error somehow?
   Because I need to use this info to update the status of my FTP server on the user interface.
   There is no exception raised, so a try/catch around the ftp server instantation does not work.

Thanks!!!

**Reply by matt-forster** [reply]
I will look to answer in full for you when I am back from holidays at the end of the week!

**Comment by bartbutenaers** [comment]
If anybody in the future should have the same questions as me: I have created a [pull request](https://github.com/QuorumDMS/ftp-srv/pull/314) to implement some extra event emitters.
